File: 22-dp1/2-memoization/main.cpp
// Memoization - C++ example

#include <iostream>
#include <unordered_map>

int fib(int n, std::unordered_map<int,int>& memo) {
    if (memo.find(n) != memo.end()) return memo[n];
    if (n <= 1) return n;
    memo[n] = fib(n-1, memo) + fib(n-2, memo);
    return memo[n];
}

int main() {
    std::unordered_map<int,int> memo;
    std::cout << fib(10, memo) << std::endl;  // Output: 55
    return 0;
}

File: 22-dp1/2-memoization/main.py
# Memoization - Python example

def fib(n, memo={}):
    if n in memo:
        return memo[n]
    if n <= 1:
        return n
    memo[n] = fib(n-1, memo) + fib(n-2, memo)
    return memo[n]

if __name__ == "__main__":
    print(fib(10))  # Output: 55
